### PR TITLE
[Federation][Kubefed] Cleanup objects on failure before exiting

### PR DESCRIPTION
**Pull request by irfanurrehman**
**What this PR does / why we need it**:
Fixes item number 3 in https://github.com/kubernetes/kubernetes/issues/41725

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
https://github.com/kubernetes/kubernetes/issues/41725

**Special notes for your reviewer**:
@madhusudancs @marun 

I would raise separate PRs for 2 and 4 but would wait for other PRs (which are in pipeline) to get merged.
2 and 4 are pretty simple to implement. I just dont want to do too many rebases, when those in pipeline get merged.

cc @perotinus 

**Release note**:

```
NONE
```

### CODE REVIEW

**Comment by irfanurrehman** [comment]
@k8s-bot verify test this
@k8s-bot kops aws e2e test this

File: federation/pkg/kubefed/init/init.go [hunk 1]
```diff
@@ -245,24 +246,24 @@ func (i *initFederation) Run(cmdOut io.Writer, config util.AdminConfig) error {
 	// 2. Expose a network endpoint for the federation API server
 	svc, ips, hostnames, err := createService(hostClientset, i.commonOptions.FederationSystemNamespace, serverName, i.options.apiServerAdvertiseAddress, i.options.apiServerServiceType, i.options.dryRun)
 	if err != nil {
-		return err
+		return cleanupObjects(hostClientset, i.commonOptions.FederationSystemNamespace, i.options.dryRun, err)
```
Comment (marun) [inline federation/pkg/kubefed/init/init.go]: I think having to call  ``cleanupObjects`` is a violation of DRY.  Consider factoring the contents of Run into a new method and have Run call that method.  If an error is returned by the method, perform cleanup.

File: federation/pkg/kubefed/unjoin.go [hunk 2]
```diff
@@ -88,7 +86,7 @@ func (u *unjoinFederation) Run(f cmdutil.Factory, cmdOut, cmdErr io.Writer, conf
 	// federation host cluster. See join_federation.go for details.
 	hostFactory := config.HostFactory(u.commonOptions.Host, u.commonOptions.Kubeconfig)
 	err = deleteSecret(hostFactory, cluster.Spec.SecretRef.Name, u.commonOptions.FederationSystemNamespace)
-	if isNotFound(err) {
+	if util.IsNotFound(err) {
```
Comment (marun) [inline federation/pkg/kubefed/unjoin.go]: (no action required) I'm not sure why the secret not being found is grounds for a warning message, but the cluster doesn't warrant the same behavior.

File: federation/pkg/kubefed/join.go [hunk 3]
```diff
@@ -155,19 +162,28 @@ func (j *joinFederation) Run(f cmdutil.Factory, cmdOut io.Writer, config util.Ad
...
 	glog.V(2).Infof("Cluster credentials secret created")
 
-	return kubectlcmd.RunCreateSubcommand(f, cmd, cmdOut, &kubectlcmd.CreateSubcommandOptions{
+	err = kubectlcmd.RunCreateSubcommand(f, cmd, cmdOut, &kubectlcmd.CreateSubcommandOptions{
 		Name:                j.commonOptions.Name,
 		StructuredGenerator: generator,
 		DryRun:              j.options.dryRun,
 		OutputFormat:        cmdutil.GetFlagString(cmd, "output"),
 	})
+	if !errors.IsAlreadyExists(err) && !j.options.dryRun {
```
Comment (marun) [inline federation/pkg/kubefed/join.go]: What happens when ``err == nil``?

File: federation/pkg/kubefed/init/init.go (hunk 1 above)
Comment (marun) [inline federation/pkg/kubefed/init/init.go]: This strategy would also allow the separation of error handling from cleanup.  Having to pass the error to ``cleanupObjects`` is a definite code smell.

File: federation/pkg/kubefed/init/init.go [hunk 4]
```diff
@@ -934,3 +939,55 @@ func updateKubeconfig(config util.AdminConfig, name, endpoint, kubeConfigPath st
 
 	return nil
 }
+
+func cleanupKubeconfig(config util.AdminConfig, name, kubeConfigPath string, dryRun bool) error {
```
Comment (marun) [inline federation/pkg/kubefed/init/init.go]: Where's the test for this?

File: federation/pkg/kubefed/init/init.go [hunk 5]
```diff
@@ -934,3 +939,55 @@ func updateKubeconfig(config util.AdminConfig, name, endpoint, kubeConfigPath st
...
+	kubeconfig, err := po.GetStartingConfig()
+	if err != nil {
+		return err
+	}
+
+	delete(kubeconfig.Clusters, name)
+	delete(kubeconfig.AuthInfos, name)
+	delete(kubeconfig.Contexts, name)
+
+	if !dryRun {
```
Comment (marun) [inline federation/pkg/kubefed/init/init.go]: Why bother passing dryRun to this function?  Why not just avoid calling it if dryRun is true?

File: federation/pkg/kubefed/init/init.go [hunk 6]
```diff
@@ -934,3 +939,55 @@ func updateKubeconfig(config util.AdminConfig, name, endpoint, kubeConfigPath st
...
+		}
+	}
+
+	return nil
+}
+
+// cleanupObjects clean up all the federation control plane objects from the host cluster
+// TODO: the load balancer resource might be outside the scope of namespace deletion on some clouds
+// to be checked and updated appropriately
+func cleanupObjects(clientset *client.Clientset, namespace string, dryRun bool, onErr error) error {
```
Comment (marun) [inline federation/pkg/kubefed/init/init.go]: Where's the test for this?

Same comment regarding dryRun.

**Comment by quinton-hoole** [comment]
@irfanurrehman This needs a rebase.

**Pull request closed without merging**